### PR TITLE
1351370: Ensure rhsmd exits on exceptions

### PR DESCRIPTION
**Pull request by cnsnyder**
The initial issue, as reported in the bz, has to do with the default selinux policy of the system. This policy prevents rhsmd from having access to files in /tmp. The check made for the existence of the directory '/tmp/product-default' on initialization of a new instance of ProductCertificateDirectory returns false due to the lack of permissions. This, in turn, causes the ProductCertificateDirectory instance to attempt to create it. The inevitably fails as in the bz above, the directory does in fact exist.

The issue that this PR actually fixes is not the one described above. The problem fixed here is that after a failure as described above rhsmd continues to run and hold on to objects that are in a stale state. This is the result of some of our injection code.

The solution was to ensure that rhsmd exits even on exceptions. An alternative would be to modify our injection code to keep creating new instances of objects when requested by rhsmd. While this alternative would fix it in theory, I believe the fix in this PR to be the simplest and most true to the desired behaviour of rhsmd.

### CODE REVIEW

File: src/daemons/rhsm_d.py [hunk 1]
```diff
@@ -133,7 +134,28 @@ def check_if_ran_once(checker, loop):
...
+    A decorator ensuring the decorated method exits the main loop after
+    running (even if it fails with an exception)
+    :param func: the method that will be decorated. Should be a method of
+    StatusChecker
+    :return:
+    """
+    try:
+        return func(*args, **kwargs)
+    finally:
+        status_checker = args[0]  # Should be a StatusChecker instance
```
Comment (barnabycourt) [inline src/daemons/rhsm_d.py]: This should log an error if it args[0] isn't a StatusChecker instance.

**Comment by wottop** [comment]
buildbot: retest this please

**Comment by cnsnyder** [comment]
@candlepin/jenkins-bot retest this please

**Comment by awood** [comment]
Looks good.  Very tidy!

**Pull request merged**